### PR TITLE
Better handle initialization errors

### PR DESCRIPTION
**Pull request by achrysaetos**
Closes https://github.com/iTwin/viewer/issues/157

### CODE REVIEW

File: common/changes/@itwin/desktop-viewer-react/leck-initialize_2022-08-02-22-04.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/desktop-viewer-react",
+      "comment": "Fixes WebInitializer.initialized error if IModelApp is already initialized",
```
Comment (aruniverse) [inline common/changes/@itwin/desktop-viewer-react/leck-initialize_2022-08-02-22-04.json:5]: DesktopInitializer

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts [hunk 2]
```diff
@@ -107,8 +107,11 @@ export class DesktopInitializer {
           DesktopInitializer._initializing = false;
           DesktopInitializer._cancel = undefined;
         });
-    } else {
-      this._initialized = Promise.resolve();
+    } else if (this._initializing) {
```
Comment (BibekSingh09) [inline packages/modules/desktop-viewer-react/src/services/Initializer.ts:110]: What would happen if _initializing is false, we don't have any condition to handle the case - 

image

Comment (achrysaetos) [reply packages/modules/desktop-viewer-react/src/services/Initializer.ts]: good catch.

**Review by BibekSingh09 — APPROVED** [review]
I have tested, this fix seems to be working for me.

**Review by kckst8 — CHANGES_REQUESTED** [review]
The thing that is missing here is another else if before throwing the error

`else if (this._initializing) {
    return this._initialized;
}`

Basically if another initialization already started we want to return the promise if that makes sense. Another way might be to check IModelApp.initialized and just throw at the beginning and then you just have 2 conditions in the conditional.

Basically there are 3 cases that need to be handled:

1. Nothing is initialized, proceed
2. IModelApp is not yet initialized but the process has been started, don't start a second initialization but also dont throw. Just let the first complete.
3. IModelApp already initialized. Throw an error directing the user to the hook instead.

**Comment by achrysaetos** [comment]
@kckst8 Makes sense! tysm

**Review by kckst8 — APPROVED** [review]

**Comment by BeAliAslam** [comment]
There is a problem in version 3.0.7 due to recent changes (via this PR) in src/services/initializer.ts>> startWebViewer function. 
The if (IModelApp.initialize) would fail if we re-render viewer from host by changing some viewerprop as useWebViewerInitializer.tsx useeffect gets kicked in and calls WebInitializer.startWebViewer. 

One use case i have is that if you add say another 'additional rpc interface' via additionalRpcInterfaces prop of viewer in host by say adding the interface in onIModelAppInit callback, the viewer throws error in startWebViewer function. 

In sample app, this can be reproduced below

App.tsx
   const [additionalRpcInterfaces, setAdditionalRpcInterfaces] = useState<RpcInterfaceDefinition<RpcInterface>[]>([]);
...
const onIModelConnected = ()=>{
...
const additionalRpcsLocal = [...additionalRpcInterfaces];

  additionalRpcsLocal.push(DevToolsRpcInterface);

setAdditionalRpcInterfaces([...additionalRpcsLocal])
}

<Viewer
        iTwinId={iTwinId ?? ""}
      ...
        additionalRpcInterfaces={additionalRpcInterfaces}
        onIModelAppInit={onIModelAppInit}
...

@achrysaetos @DarrinDoherty

**Comment by aruniverse** [comment]
@BeAliAslam , ive opened a pr to revert this change for the time being, https://github.com/iTwin/viewer/pull/217

**Pull request merged**